File: uci-python-v2/uci/core/registry.py
"""
UCI Registry
In-memory node registry — tracks all known nodes, their manifests,
trust records, and mounted capabilities.

Registries are advisory only. They MUST NOT override governance policy.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional
from .manifest import UCIManifest
from .trust import TrustRecord, TrustState
from .errors import UCIRegistryError


@dataclass
class NodeEntry:
    """A single registered node and everything the registry knows about it."""
    node_id: str
    manifest: UCIManifest
    trust: TrustRecord
    registered_at: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    last_seen: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    mounted_capabilities: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)

    def touch(self) -> None:
        self.last_seen = datetime.now(timezone.utc).isoformat()

    def is_ready(self) -> bool:
        return (
            self.trust.can_execute()
            and bool(self.mounted_capabilities)
        )


@dataclass
class UCIRegistry:
    """
    Central index of all UCI nodes the orchestrator has encountered.
    Thread-safety is the caller's responsibility in this v0.1 implementation.
    """
    _nodes: dict[str, NodeEntry] = field(default_factory=dict, repr=False)

    # ── Registration ─────────────────────────────────

    def register(
        self,
        manifest: UCIManifest,
        trust: TrustRecord,
        tags: Optional[list[str]] = None,
    ) -> NodeEntry:
        node_id = manifest.node.node_id
        if node_id in self._nodes:
            raise UCIRegistryError(
                f"Node '{node_id}' is already registered. Use update() to refresh."
            )
        entry = NodeEntry(
            node_id    = node_id,
            manifest   = manifest,
            trust      = trust,
            tags       = tags or [],
        )
        self._nodes[node_id] = entry
        return entry

    def update(self, manifest: UCIManifest, trust: TrustRecord) -> NodeEntry:
        """Re-register an existing node (e.g. after revalidation)."""
        node_id = manifest.node.node_id
        if node_id not in self._nodes:
            return self.register(manifest, trust)
        entry = self._nodes[node_id]
        entry.manifest = manifest
        entry.trust    = trust
        entry.touch()
        return entry

    def deregister(self, node_id: str) -> None:
        if node_id not in self._nodes:
            raise UCIRegistryError(f"Node '{node_id}' not found in registry.")
        del self._nodes[node_id]

    # ── Capability mounting ──────────────────────────

    def mount_capability(self, node_id: str, capability_id: str) -> None:
        entry = self._get(node_id)
        if capability_id not in entry.mounted_capabilities:
            entry.mounted_capabilities.append(capability_id)

    def revoke_capability(self, node_id: str, capability_id: str) -> None:
        entry = self._get(node_id)
        entry.mounted_capabilities = [
            c for c in entry.mounted_capabilities if c != capability_id
        ]

    # ── Lookups ──────────────────────────────────────

    def get(self, node_id: str) -> Optional[NodeEntry]:
        return self._nodes.get(node_id)

    def require(self, node_id: str) -> NodeEntry:
        entry = self.get(node_id)
        if entry is None:
            raise UCIRegistryError(f"Node '{node_id}' not found in registry.")
        return entry

    def all_nodes(self) -> list[NodeEntry]:
        return list(self._nodes.values())

    def ready_nodes(self) -> list[NodeEntry]:
        return [e for e in self._nodes.values() if e.is_ready()]

    def nodes_with_capability(self, capability_id: str) -> list[NodeEntry]:
        return [
            e for e in self._nodes.values()
            if capability_id in e.mounted_capabilities
        ]

    def nodes_by_trust(self, state: TrustState) -> list[NodeEntry]:
        return [e for e in self._nodes.values() if e.trust.state == state]

    def count(self) -> int:
        return len(self._nodes)

    def _get(self, node_id: str) -> NodeEntry:
        entry = self._nodes.get(node_id)
        if entry is None:
            raise UCIRegistryError(f"Node '{node_id}' not found.")
        return entry

    # ── Summary ──────────────────────────────────────

    def summary(self) -> dict:
        return {
            "total_nodes":   self.count(),
            "ready":         len(self.ready_nodes()),
            "nodes": [
                {
                    "node_id":    e.node_id,
                    "trust":      e.trust.state.value,
                    "capabilities": e.mounted_capabilities,
                    "is_ready":   e.is_ready(),
                }
                for e in self._nodes.values()
            ],
        }

File: uci-typescript/src/registry.ts
/**
 * UCI Registry & Handshake
 */

import { UCIManifest }    from "./manifest.js";
import { TrustRecord, TrustState } from "./trust.js";
import { UCIRegistryError }        from "./errors.js";
import { AuditLog, AuditEvent }    from "./audit.js";
import { PolicyEngine, GovernanceOutcome } from "./governance.js";

// ── Registry ────────────────────────────────────────────────

export interface NodeEntry {
  nodeId:               string;
  manifest:             UCIManifest;
  trust:                TrustRecord;
  registeredAt:         string;
  mountedCapabilities:  string[];
}

export class UCIRegistry {
  private readonly _nodes = new Map<string, NodeEntry>();

  register(manifest: UCIManifest, trust: TrustRecord): NodeEntry {
    const nodeId = manifest.nodeId;
    if (this._nodes.has(nodeId)) {
      throw new UCIRegistryError(`Node '${nodeId}' is already registered.`);
    }
    const entry: NodeEntry = {
      nodeId,
      manifest,
      trust,
      registeredAt:        new Date().toISOString(),
      mountedCapabilities: [],
    };
    this._nodes.set(nodeId, entry);
    return entry;
  }

  update(manifest: UCIManifest, trust: TrustRecord): NodeEntry {
    const nodeId = manifest.nodeId;
    if (!this._nodes.has(nodeId)) return this.register(manifest, trust);
    const entry = this._nodes.get(nodeId)!;
    entry.manifest = manifest;
    entry.trust    = trust;
    return entry;
  }

  get(nodeId: string):     NodeEntry | undefined { return this._nodes.get(nodeId); }
  require(nodeId: string): NodeEntry {
    const e = this._nodes.get(nodeId);
    if (!e) throw new UCIRegistryError(`Node '${nodeId}' not found in registry.`);
    return e;
  }

  mountCapability(nodeId: string, capabilityId: string): void {
    const e = this.require(nodeId);
    if (!e.mountedCapabilities.includes(capabilityId)) {
      e.mountedCapabilities.push(capabilityId);
    }
  }

  allNodes():   NodeEntry[] { return [...this._nodes.values()]; }
  readyNodes(): NodeEntry[] {
    return this.allNodes().filter(
      e => e.trust.canExecute() && e.mountedCapabilities.length > 0
    );
  }

  count(): number { return this._nodes.size; }

  summary() {
    return {
      total_nodes: this.count(),
      ready:       this.readyNodes().length,
      nodes:       this.allNodes().map(e => ({
        node_id:      e.nodeId,
        trust:        e.trust.state,
        capabilities: e.mountedCapabilities,
        is_ready:     e.trust.canExecute() && e.mountedCapabilities.length > 0,
      })),
    };
  }
}

// ── Handshake ───────────────────────────────────────────────

export const HandshakeStage = {
  PENDING:               "pending",
  DISCOVERED:            "discovered",
  MANIFEST_RETRIEVED:    "manifest_retrieved",
  MANIFEST_VALIDATED:    "manifest_validated",
  COMPATIBILITY_CHECKED: "compatibility_checked",
  GOVERNANCE_EVALUATED:  "governance_evaluated",
  TRUST_ASSIGNED:        "trust_assigned",
  CAPABILITIES_MOUNTED:  "capabilities_mounted",
  READY:                 "ready",
  FAILED:                "failed",
} as const;

export type HandshakeStageValue = typeof HandshakeStage[keyof typeof HandshakeStage];

export interface HandshakeResult {
  success:              boolean;
  nodeId:               string;
  stageReached:         HandshakeStageValue;
  trustState:           string;
  mountedCapabilities:  string[];
  failureReason:        string;
  warnings:             string[];
}

const SUPPORTED_VERSIONS = ["0.1"];

export class HandshakeEngine {
  constructor(
    private readonly policy:   PolicyEngine,
    private readonly registry: UCIRegistry,
    private readonly audit:    AuditLog,
  ) {}

  run(nodeId: string, manifestData: unknown, preTrusted = false): HandshakeResult {
    let   stage    = HandshakeStage.PENDING as HandshakeStageValue;
    const trust    = new TrustRecord(nodeId);
    const warnings: string[] = [];

    const fail = (reason: string): HandshakeResult => {
      this.audit.append({
        eventType: AuditEvent.HANDSHAKE_FAILED, nodeId,
        actor: "handshake_engine", outcome: "deny",
        detail: { stage, reason },
      });
      return { success: false, nodeId, stageReached: stage,
               trustState: trust.state, mountedCapabilities: [],
               failureReason: reason, warnings };
    };

    try {
      // Stage 1: Discovered
      stage = HandshakeStage.DISCOVERED;
      trust.transition(TrustState.DISCOVERED, "handshake_engine", "Node endpoint reached");
      this.audit.append({ eventType: AuditEvent.NODE_DISCOVERED, nodeId, actor: "handshake_engine" });

      // Stage 2: Manifest retrieved
      stage = HandshakeStage.MANIFEST_RETRIEVED;
      this.audit.append({ eventType: AuditEvent.MANIFEST_RETRIEVED, nodeId, actor: "handshake_engine" });

      // Stage 3: Manifest validated
      stage = HandshakeStage.MANIFEST_VALIDATED;
      let manifest: UCIManifest;
      try {
        manifest = UCIManifest.fromDict(manifestData);
        manifest.validate();
      } catch (err) {
        this.audit.append({
          eventType: AuditEvent.MANIFEST_VALIDATION_FAILED, nodeId,
          actor: "handshake_engine", outcome: "deny",
          detail: { reason: String(err) },
        });
        return fail(`Manifest validation failed: ${err}`);
      }
      this.audit.append({
        eventType: AuditEvent.MANIFEST_VALIDATION_PASSED, nodeId,
        actor: "handshake_engine",
        detail: { capabilities: manifest.capabilityIds() },
      });

      // Stage 4: Compatibility
      stage = HandshakeStage.COMPATIBILITY_CHECKED;
      if (!manifest.isCompatibleWith(SUPPORTED_VERSIONS)) {
        this.audit.append({
          eventType: AuditEvent.COMPATIBILITY_REJECTED, nodeId,
          actor: "handshake_engine", outcome: "deny",
          detail: { manifest_version: manifest.version },
        });
        return fail(`Manifest version '${manifest.version}' not in supported ${SUPPORTED_VERSIONS}.`);
      }
      this.audit.append({ eventType: AuditEvent.COMPATIBILITY_ACCEPTED, nodeId, actor: "handshake_engine" });
      trust.transition(TrustState.VERIFIED, "handshake_engine", "Manifest validated and compatible");

      // Stage 5: Governance
      stage = HandshakeStage.GOVERNANCE_EVALUATED;
      const govDecision = this.policy.evaluateManifest(manifest, trust);
      this.audit.append({
        eventType: AuditEvent.POLICY_EVALUATED, nodeId,
        actor: "policy_engine", outcome: govDecision.outcome,
        detail: { reason: govDecision.reason },
      });
      if (govDecision.outcome === GovernanceOutcome.DENY) {
        return fail(`Governance denied: ${govDecision.reason}`);
      }

      // Stage 6: Trust assigned
      stage = HandshakeStage.TRUST_ASSIGNED;
      const newTrust = (preTrusted || govDecision.outcome === GovernanceOutcome.ALLOW)
        ? TrustState.TRUSTED
        : TrustState.RESTRICTED;
      trust.transition(newTrust, "policy_engine", govDecision.reason);
      this.audit.append({
        eventType: AuditEvent.TRUST_ASSIGNED, nodeId,
        actor: "policy_engine",
        detail: { trust_state: newTrust },
      });

      // Stage 7: Mount capabilities
      stage = HandshakeStage.CAPABILITIES_MOUNTED;
      const mounted: string[] = [];
      try { this.registry.register(manifest, trust); }
      catch { this.registry.update(manifest, trust); }

      for (const cap of manifest.data.capabilities) {
        if (newTrust === TrustState.RESTRICTED) {
          const hasHighRisk = cap.actions.some(
            a => ["high", "critical"].includes(a.risk?.level ?? "")
          );
          if (hasHighRisk) {
            warnings.push(
              `Capability '${cap.capability_id}' not mounted — contains high-risk actions and node is restricted.`
            );
            this.audit.append({
              eventType: AuditEvent.CAPABILITY_REVOKED, nodeId,
              actor: "handshake_engine", outcome: "deny",
              detail: { capability_id: cap.capability_id },
            });
            continue;
          }
        }
        this.registry.mountCapability(nodeId, cap.capability_id);
        mounted.push(cap.capability_id);
        this.audit.append({
          eventType: AuditEvent.CAPABILITY_MOUNTED, nodeId,
          actor: "handshake_engine",
          detail: { capability_id: cap.capability_id },
        });
      }

      // Stage 8: Ready
      stage = HandshakeStage.READY;
      this.audit.append({
        eventType: AuditEvent.NODE_READY, nodeId,
        actor: "handshake_engine",
        detail: { mounted_capabilities: mounted, trust_state: newTrust },
      });

      return {
        success:             true,
        nodeId,
        stageReached:        HandshakeStage.READY,
        trustState:          newTrust,
        mountedCapabilities: mounted,
        failureReason:       "",
        warnings,
      };

    } catch (err) {
      return fail(`Unexpected failure at stage '${stage}': ${err}`);
    }
  }
}
